src/m365/teams/commands/message/message-export.ts

```diff
@@ -0,0 +1,218 @@
+          return `${args.options.userId} is not a valid GUID`;
+        }
+
+        if (args.options.userName && !validation.isValidUserPrincipalName(args.options.userName)) {
+          return `${args.options.userName} is not a valid userPrincipalName`;
+        }
+
+        if (args.options.teamId && !validation.isValidGuid(args.options.teamId)) {
+          return `${args.options.teamId} is not a valid GUID`;
+        }
+
+        if (args.options.fromDateTime && !validation.isValidISODateTime(args.options.fromDateTime)) {
+          return `${args.options.fromDateTime} is not a valid ISO DateTime`;
+        }
+
+        if (args.options.toDateTime && !validation.isValidISODateTime(args.options.toDateTime)) {
+          return `${args.options.toDateTime} is not a valid ISO DateTime`;
+        }
+
+        if (args.options.folderPath && !fs.existsSync(args.options.folderPath)) {
+          return `Path ${args.options.folderPath} does not exist.`;
+        }
+
+        return true;
+      }
+    );
+  }
+
+  #initOptionSets(): void {
+    this.optionSets.push(
+      {
+        options: ['userId', 'userName', 'teamId']
+      },
+      {
+        options: ['folderPath'],
+        runsWhen: (args: CommandArgs) => {
+          return args.options.withAttachments !== undefined && args.options.withAttachments;
+        }
+      },
+      {
+        options: ['withAttachments'],
+        runsWhen: (args: CommandArgs) => {
+          return args.options.folderPath !== undefined;
+        }
+      }
+    );
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    let baseUrl = `${this.resource}/v1.0/`;
+
+    if (args.options.userId || args.options.userName) {
+      baseUrl += `users/${args.options.userId || args.options.userName}/chats`;
+    }
+    else {
+      baseUrl += `teams/${args.options.teamId}/channels`;
+    }
+
+    let requestUrl = `${baseUrl}/getAllMessages`;
+
+    const filters = this.getFilters(args.options);
+    if (filters.length > 0) {
+      requestUrl += `?$filter=${filters.join(' and ')}`;
+    }
+
+    try {
+      const res: any[] = await odata.getAllItems(requestUrl);
+      if (args.options.withAttachments) {
+        for (const message of res) {
+          if (message.attachments && message.attachments.length > 0) {
+            for await (const attachment of message.attachments) {
+              const _url = url.parse(attachment['contentUrl']);
+              let siteUrl = _url.protocol + '//' + _url.host!;
+
+              if (_url.path!.split('/')[1] === 'sites' || _url.path!.split('/')[1] === 'teams' || _url.path!.split('/')[1] === 'personal') {
+                siteUrl += '/' + _url.path!.split('/')[1] + '/' + _url.path!.split('/')[2];
+              }
+
+              const requestOptions: CliRequestOptions = {
+                url: `${siteUrl}/_api/web/getFileByServerRelativePath(decodedUrl='${_url.path!}')/$value`,
```

Comment:
We may need to encode it first. That's what we do in other locations as well. Otherwise it might break when special chars are in the url.

Also: for path length limitations, could you use a url variable here?

`(DecodedUrl=@decodedUrl)?@decodedUrl='<url>'`

Reply:
Hi @martinlingstuyl 

When exporting messages, the URL is already encoded, so if I encode it another time, it will not work it seems.